setup.cfg

```diff
@@ -93,6 +93,9 @@ all =
     ipython>=4.2
     pytest>=7.0
     typing_extensions>=3.10.0.1
+    fsspec
+    s3fs
+    gcsfs
```

Comment:
We could definitely create a special `[cloud]` category for these if we are worried about bloating the dependencies.

Some background details:
* `s3fs` and `gcsfs` are both fairly lightweight pure Python packages which enable logic specific to AWS or Google Cloud to be left out of the `fsspec` core package.
* Including `s3fs` made sense to me because the data archives of several NASA missions are hosted in S3 storage.
* We could remove `gcsfs` because I am not aware of any astronomy projects which use Google Cloud storage right now, but I included it to avoid giving the impression that we prefer one vendor over another.

Reply:
> astronomy projects which use Google Cloud storage

rubin?

Reply:
Since `fsspec` supports many different services (see below) and many those of those require some optional dependency, and since `fsspec` will tell the user if a dependency is missing, I think we should just have an optional dependency on `fsspec` itself and delegate to it to warn the user if some toher package is needed.

<details>
```
In [5]: known_implementations
Out[5]: 
{'abfs': {'class': 'adlfs.AzureBlobFileSystem',
          'err': 'Install adlfs to access Azure Datalake Gen2 and Azure Blob '
                 'Storage'},
 'adl': {'class': 'adlfs.AzureDatalakeFileSystem',
         'err': 'Install adlfs to access Azure Datalake Gen1'},
 'arrow_hdfs': {'class': 'fsspec.implementations.arrow.HadoopFileSystem',
                'err': 'pyarrow and local java libraries required for HDFS'},
 'az': {'class': 'adlfs.AzureBlobFileSystem',
        'err': 'Install adlfs to access Azure Datalake Gen2 and Azure Blob '
               'Storage'},
 'blockcache': {'class': 'fsspec.implementations.cached.CachingFileSystem'},
 'cached': {'class': 'fsspec.implementations.cached.CachingFileSystem'},
 'dask': {'class': 'fsspec.implementations.dask.DaskWorkerFileSystem',
          'err': 'Install dask distributed to access worker file system'},
 'dbfs': {'class': 'fsspec.implementations.dbfs.DatabricksFileSystem',
          'err': 'Install the requests package to use the '
                 'DatabricksFileSystem'},
 'dropbox': {'class': 'dropboxdrivefs.DropboxDriveFileSystem',
             'err': 'DropboxFileSystem requires "dropboxdrivefs","requests" '
                    'and "dropbox" to be installed'},
 'file': {'class': 'fsspec.implementations.local.LocalFileSystem'},
 'filecache': {'class': 'fsspec.implementations.cached.WholeFileCacheFileSystem'},
 'ftp': {'class': 'fsspec.implementations.ftp.FTPFileSystem'},
 'gcs': {'class': 'gcsfs.GCSFileSystem',
         'err': 'Please install gcsfs to access Google Storage'},
 'gdrive': {'class': 'gdrivefs.GoogleDriveFileSystem',
            'err': 'Please install gdrivefs for access to Google Drive'},
 'git': {'class': 'fsspec.implementations.git.GitFileSystem',
         'err': 'Install pygit2 to browse local git repos'},
 'github': {'class': 'fsspec.implementations.github.GithubFileSystem',
            'err': 'Install the requests package to use the github FS'},
 'gs': {'class': 'gcsfs.GCSFileSystem',
        'err': 'Please install gcsfs to access Google Storage'},
 'hdfs': {'class': 'fsspec.implementations.hdfs.PyArrowHDFS',
          'err': 'pyarrow and local java libraries required for HDFS'},
 'http': {'class': 'fsspec.implementations.http.HTTPFileSystem',
          'err': 'HTTPFileSystem requires "requests" and "aiohttp" to be '
                 'installed'},
 'https': {'class': 'fsspec.implementations.http.HTTPFileSystem',
           'err': 'HTTPFileSystem requires "requests" and "aiohttp" to be '
                  'installed'},
 'jlab': {'class': 'fsspec.implementations.jupyter.JupyterFileSystem',
          'err': 'Jupyter FS requires requests to be installed'},
 'jupyter': {'class': 'fsspec.implementations.jupyter.JupyterFileSystem',
             'err': 'Jupyter FS requires requests to be installed'},
 'libarchive': {'class': 'fsspec.implementations.libarchive.LibArchiveFileSystem',
                'err': 'LibArchive requires to be installed'},
 'memory': {'class': 'fsspec.implementations.memory.MemoryFileSystem'},
 'oci': {'class': 'ocifs.OCIFileSystem',
         'err': 'Install ocifs to access OCI Object Storage'},
 'reference': {'class': 'fsspec.implementations.reference.ReferenceFileSystem'},
 's3': {'class': 's3fs.S3FileSystem', 'err': 'Install s3fs to access S3'},
 's3a': {'class': 's3fs.S3FileSystem', 'err': 'Install s3fs to access S3'},
 'sftp': {'class': 'fsspec.implementations.sftp.SFTPFileSystem',
          'err': 'SFTPFileSystem requires "paramiko" to be installed'},
 'simplecache': {'class': 'fsspec.implementations.cached.SimpleCacheFileSystem'},
 'smb': {'class': 'fsspec.implementations.smb.SMBFileSystem',
         'err': 'SMB requires "smbprotocol" or "smbprotocol[kerberos]" '
                'installed'},
 'ssh': {'class': 'fsspec.implementations.sftp.SFTPFileSystem',
         'err': 'SFTPFileSystem requires "paramiko" to be installed'},
 'tar': {'class': 'fsspec.implementations.tar.TarFileSystem'},
 'wandb': {'class': 'wandbfs.WandbFS',
           'err': 'Install wandbfs to access wandb'},
 'webhdfs': {'class': 'fsspec.implementations.webhdfs.WebHDFS',
             'err': 'webHDFS access requires "requests" to be installed'},
 'zip': {'class': 'fsspec.implementations.zip.ZipFileSystem'}}
```
</details>

Reply:
Agree with this; except that since fsspec itself is so small and with no deps of it's own, you may well want to include it as required and not need to check for it.

Reply:
Hmm I see the interest of doing that but Astropy always had a very limited number of required dependencies and probably more important, the code in `io.fits` dealing with the file descriptors is already quite complicated and I'm not sure I want to have another layer dealing with the fd.

Reply:
These are good points!  There is no doubt that adding dependencies to Astropy (optional or otherwise) risks creating work for maintainers in the future, so it's definitely important to trade this risk against the benefits.

My two cents:
* Adding the core `fsspec` package as an optional dependency is relatively low risk, because it is a pure Python package, has zero dependencies of its own, and is already an optional dependency of popular packages such as `pandas` and `dask`.  The core package would enable FITS files to be opened efficiently via the `http` protocol.
* There is a case to be made for including `s3fs` alongside `fsspec`, because NASA already hosts Hubble/Kepler/TESS/Chandra data in S3, and it appears to be moving towards adopting S3 for many other science missions (can anyone confirm exact plans?).  It is true however that the adoption of S3 in astronomy isn't exactly ubiquitous right now, and `s3fs` does have two dependencies of its own (`aiobotocore` and `aiohttp`), so I'd be OK with waiting to adopt `s3fs` as an optional dependency until S3 gains more significant adoption.
* The case for `gcsfs` is weak because I am not aware of any facilities currently providing data via Google Storage.  Rubin Observatory (LSST) appears to rely on Google Cloud, but someone would have to confirm whether or not the observatory plans to enable users to access FITS files via `gs://` paths.  Until then, I'd be OK with skipping `gcsfs` as an optional dependency.

Any thoughts?

Reply:
Unfortunately, fsspec does also require aiohttp even for http. The principal advantage from a user's point of view, is that there would be a concrete message saying which additional dependencies should be installed, when trying to access a storage backend not currently available.

Of the protocols, s3 may well be the most important, partly because other services exist outside of AWS which support the same API. However, I have been surprised by the amount of data funnelling to Azure (abfs), albeit not FITS.